numpyro/hmc_util.py

```diff
@@ -781,3 +782,81 @@ def single_chain_init(key, only_params=False):
         _, potential_fn, constrain_fun = single_chain_init(rng[0])
         init_params = vmap(lambda rng: single_chain_init(rng, only_params=True))(rng)
         return init_params, potential_fn, constrain_fun
+
+
+# TODO: to be replaced by np.cov ddof=1 for the next JAX version
+def _cov(samples):
+    wc_init, wc_update, wc_final = welford_covariance(diagonal=False)
+    state = wc_init(samples.shape[1])
+    state = fori_loop(0, samples.shape[0], lambda i, state: wc_update(samples[i], state), state)
+    return wc_final(state)[0]
+
+
+def consensus(subposteriors, num_draws, rng=None):
+    """
+    Merges subposteriors following consensus Monte Carlo algorithm.
+
+    **References:**
+
+    1. *Bayes and big data: The consensus Monte Carlo algorithm*,
+       Steven L. Scott, Alexander W. Blocker, Fernando V. Bonassi, Hugh A. Chipman,
+       Edward I. George, Robert E. McCulloch
+
+    :param list subposteriors: a list which each element is a collection of samples.
+    :param int num_draws: number of draws from the merged posterior.
+    :param jax.random.PRNGKey rng: source of the randomness, defaults to `jax.random.PRNGKey(0)`.
+    :return: a collection of `num_draws` samples with the same data structure as each subposterior.
+    """
+    rng = random.PRNGKey(0) if rng is None else rng
+    # stack subposteriors
+    joined_subposteriors = tree_multimap(lambda *args: np.stack(args), *subposteriors)
+    # shape = num_subposteriors x num_samples x sample_shape
+    joined_subposteriors = vmap(vmap(lambda sample: ravel_pytree(sample)[0]))(joined_subposteriors)
+
+    # randomly gets num_draws from subposteriors
+    n_subs = len(subposteriors)
+    n_samples = tree_flatten(subposteriors[0])[0][0].shape[0]
+    draw_idxs = random.randint(rng, shape=(n_subs, num_draws), minval=0, maxval=n_samples)
```

Comment:
what's going on here can you please add more comments?

Reply:
This is to "randomly gets num_draws from subposteriors" (with replacement). For each subposterior, we get `num_draws` draws from it.

Reply:
I just add a comment `# shape of draw_idxs: n_subs x num_draws x sample_shape`. Hope that it can explain the code.